### PR TITLE
Model Defination Of EfficientNet.

### PR DESCRIPTION
**Pull request by sufeidechabei**
The model defination of efficientnet. I'm not sure whether the model can achieve the result of the original paper, training from scratch on imagenet is too slow, I don't have enough GPU, may be someone can help me tune it.  Training files are in my repo (https://github.com/sufeidechabei/EfficientNet-Gluon).

### CODE REVIEW

File: gluoncv/model_zoo/model_zoo.py [hunk 1]
```diff
@@ -211,6 +212,7 @@
     'ssd_512_mobilenet1.0_voc_int8': ssd_512_mobilenet1_0_voc_int8,
     'ssd_512_resnet50_v1_voc_int8': ssd_512_resnet50_v1_voc_int8,
     'ssd_512_vgg16_atrous_voc_int8': ssd_512_vgg16_atrous_voc_int8,
+    'efficientnet': efficientnet
```
Comment (hetong007) [inline gluoncv/model_zoo/model_zoo.py]: please add b0 to b7 separately.

**Comment by mli** [comment]
Job PR-814-25 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-814/25/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/efficientnet.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+    }
+    width_coeff, depth_coeff, input_resolution, dropout_rate = params_dict[model_name]
+    blocks_args, global_params = efficientnet_param(width_coeff, depth_coeff, dropout_rate)
+    model = EfficientNet(blocks_args, global_params)
+    if return_input_resolution:
+        return model, input_resolution
+    return model
+
+
+def efficientnet_b0(pretrained=False):
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: At least, users need to be able to customize the number of classes.

Comment (hetong007) [reply gluoncv/model_zoo/efficientnet.py]: And also `ctx`. Please use the style as https://github.com/dmlc/gluon-cv/blob/master/gluoncv/model_zoo/resnetv1b.py#L340

File: gluoncv/model_zoo/efficientnet.py [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    """ Creates a efficientnet model. """
+    blocks_args = [
+        'r1_k3_s11_e1_i32_o16_se0.25', 'r2_k3_s22_e6_i16_o24_se0.25',
+        'r2_k5_s22_e6_i24_o40_se0.25', 'r3_k3_s22_e6_i40_o80_se0.25',
+        'r3_k5_s11_e6_i80_o112_se0.25', 'r4_k5_s22_e6_i112_o192_se0.25',
+        'r1_k3_s11_e6_i192_o320_se0.25',
+    ]
+    blocks_args = BlockDecoder.decode(blocks_args)
+
+    global_params = GlobalParams(
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: Don't hide options in functions, users need to be able to make modification.

File: gluoncv/model_zoo/efficientnet.py [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+                _add_conv(
+                    self._conv_head,
+                    out_channels,
+                    active=True,
+                    batchnorm=True)
+            # Final linear layer
+            self._dropout = self._global_params.dropout_rate
+            self._fc = nn.Dense(self._global_params.num_classes)
+
+    def hybrid_forward(self, F, x, training):
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: why do we need an additional "training" here?

**Comment by mli** [comment]
Job PR-814-32 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-814/32/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-814-31 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-814/31/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/efficientnet.py (hunk 4 above)
Comment (hetong007) [reply gluoncv/model_zoo/efficientnet.py]: By using `autograd.is_training()` , the implementation can get rid of the `training` argument.

File: gluoncv/model_zoo/efficientnet.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+            groups=num_group,
+            use_bias=False))
+    if batchnorm:
+        out.add(nn.BatchNorm(scale=True, momentum=0.99, epsilon=1e-3))
+    if active:
+        out.add(nn.Swish())
+
+
+class MBConv(nn.HybridBlock):
+    def __init__(
```
[the remark this answers is not in the document]
Comment (hetong007) [reply gluoncv/model_zoo/efficientnet.py]: fix the format.

File: gluoncv/model_zoo/efficientnet.py [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+        oh, ow = math.ceil(ih / sh), math.ceil(iw / sw)
+        pad_h = max((oh - 1) * self.stride[0] +
+                    (kh - 1) * self.dilation[0] + 1 - ih, 0)
+        pad_w = max((ow - 1) * self.stride[1] +
+                    (kw - 1) * self.dilation[1] + 1 - iw, 0)
+        if pad_h > 0 or pad_w > 0:
+            x = F.pad(
+                x,
+                mode='constant',
+                pad_width=(
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: fix the format.

File: gluoncv/model_zoo/efficientnet.py [hunk 7]
```diff
@@ -0,0 +1,850 @@
...
+                    2))
+            return x
+        return x
+
+
+def _add_conv(out, channels=1, kernel=1, stride=1, pad=0,
+              num_group=1, active=True, batchnorm=True):
+    out.add(SamePadding(kernel, stride, dilation=(1, 1)))
+    out.add(
+        nn.Conv2D(
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: fix the format.

File: gluoncv/model_zoo/efficientnet.py [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+            self._fc = nn.Dense(num_classes)
+
+    def hybrid_forward(self, F, x, training):
+        x = self.features(x)
+        for block in self._blocks:
+            block.get_mode(training=training)
+            x = block(x)
+        x = self._conv_head(x)
+        x = F.squeeze(
+            F.squeeze(
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: fix the format.

**Comment by mli** [comment]
Job PR-814-35 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-814/35/index.html 
Code coverage of this PR:  vs. Master:

**Comment by sufeidechabei** [comment]
dong @hetong007

File: gluoncv/model_zoo/efficientnet.py [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+        if isinstance(kernel_size, int):
+            kernel_size = (kernel_size,) * 2
+        if isinstance(stride, int):
+            stride = (stride,) * 2
+        self.kernel_size = kernel_size
+        self.stride = stride
+        self.dilation = dilation
+
+    def hybrid_forward(self, F, x):
+        ih, iw = x.shape[-2:]
```
Comment (hetong007) [inline gluoncv/model_zoo/efficientnet.py]: Why do we need use the shape here?

**Comment by hetong007** [comment]
Please use the following command to check your implementation:

```
python train_imagenet.py \
  --rec-train /media/ramdisk/rec/train.rec --rec-train-idx /media/ramdisk/rec/train.idx \
  --rec-val /media/ramdisk/rec/val.rec --rec-val-idx /media/ramdisk/rec/val.idx \
  --model efficientnet_b0 --mode hybrid_ \
  --lr 0.4 --lr-mode cosine --num-epochs 120 --batch-size 128 --num-gpus 0 -j 60 \
  --warmup-epochs 5 --dtype float16 \
  --use-rec --no-wd --label-smoothing \
  --save-dir params_efficientnet_b0 \
  --logging-file efficientnet_b0.log
```

Currently it has some issues blocking the training.

File: gluoncv/model_zoo/efficientnet.py (hunk 9 above)
Comment (zhanghang1989) [reply gluoncv/model_zoo/efficientnet.py]: Reading shape in hybrid will fail. Please use `Block` install of `HybridBlock`

Comment (sufeidechabei) [reply gluoncv/model_zoo/efficientnet.py]: okay, I will change the code@zhanghang1989

**Pull request closed without merging**